partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/FnRegistry.kt

```diff
@@ -0,0 +1,95 @@
+package org.partiql.planner.internal.typer
+
+import org.partiql.planner.internal.ir.Agg
+import org.partiql.planner.internal.ir.Fn
+import org.partiql.planner.internal.ir.Identifier
+import org.partiql.planner.internal.typer.FnBuiltins.pAggs
+import org.partiql.planner.internal.typer.FnBuiltins.pCasts
+import org.partiql.planner.internal.typer.FnBuiltins.pFns
+import org.partiql.planner.internal.typer.FnBuiltins.pOps
+import org.partiql.spi.connector.ConnectorFunctions
+import org.partiql.types.function.FunctionSignature
+import org.partiql.value.PartiQLValueExperimental
+import org.partiql.value.PartiQLValueType
+
+/**
+ * Function signature lookup by name.
+ */
+internal typealias FnMap<T> = Map<String, List<T>>
+
+/**
+ * This class is responsible for quickly looking up function signatures given a function name. We will expand upon
+ * this to support the resolution of function names via the SQL path once ready.
+ *
+ * Note that this automatically includes the PartiQLMetadata as it would be wasteful to reproduce this every time.
+ */
+@PartiQLValueExperimental
+internal class FnRegistry(private val metadata: Collection<ConnectorFunctions>) {
+
+    /**
+     * Calculate a queryable map of scalar function signatures.
+     */
+    private val uFns: FnMap<FunctionSignature.Scalar> = metadata.flatMap { it.functions }.toFnMap()
+
+    /**
+     * Calculate a queryable map of scalar function signatures from special forms.
+     */
+    private val uOps: FnMap<FunctionSignature.Scalar> = metadata.flatMap { it.operators }.toFnMap()
+
+    /**
+     * Calculate a queryable map of aggregation function signatures
+     */
+    private val uAggs: FnMap<FunctionSignature.Aggregation> = metadata.flatMap { it.aggregations }.toFnMap()
```

Comment:
I feel like we can just load all functions and operators here instead of having a separate FnBuiltins. 

Also, not in the scope of this PR, but we might need a better model around user defined operators. Currently this is no way for user to load in customer operator implementation.